### ISSUE TITLE
First DataField is always wrong

### ISSUE THREAD

**Issue by mariussoutier** [nyoka-pmml/nyoka#6]
Hi,

I have a fairly straightforward sklearn Pipeline that uses a Vectorizer and gradient boost. Exporting it using Nyoka works except that the first DataField is output as a categorical string, although all fields should just be continuous doubles, the training input is a NumPy NDArray of numbers.

```
    <DataDictionary numberOfFields="1501">
        <DataField name="aa" optype="categorical" dataType="string"/>
        <DataField name="aan" optype="continuous" dataType="double"/>
        <DataField name="abs" optype="continuous" dataType="double"/>
...
```

**Comment by Nirmal-Neel** [comment]
Hi @mariussoutier, can you provide your model structure or the model summary?

**Comment by mariussoutier** [comment]
I'm not an expert on PMML, what kind of info do you need?

**Comment by sharmasw** [comment]
Hi @mariussoutier, could you please share your pipeline structure that you are trying to convert?

**Comment by mariussoutier** [comment]
```
<Extension value="Pipeline(memory=None,
    steps=[('vectorizer',CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=&lt;class 'numpy.int64'&gt;, encoding='utf-8', input='filename',
        lowercase=True, max_df=1.0, max_features=1500, min_df=1,
        ngram_range=(1, 1), preprocessor=None,
        stop_words=[...],
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)),('alg',GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.06, loss='deviance', max_depth=7,
              max_features=None, max_leaf_nodes=None,
              min_impurity_decrease=0.0, min_impurity_split=None,
              min_samples_leaf=1, min_samples_split=3,
              min_weight_fraction_leaf=0.0, n_estimators=15,
              n_iter_no_change=None, presort='auto', random_state=31337,
              subsample=1, tol=0.0001, validation_fraction=0.1, verbose=1,
              warm_start=False))])">
        </Extension>
```

**Comment by Nirmal-Neel** [comment]
Hi @mariussoutier,  could you please provide a script that reproduces the error?

**Comment by mariussoutier** [comment]
Hi @Nirmal-Neel, sure, here you go.

```
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split
from nyoka import skl_to_pmml

data = [
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc,',
    'Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles. Ma quande lingues coalesce, li grammatica del resultant lingue es plu simplic e regulari quam ti del coalescent lingues. Li nov lingua franca va esser plu simplic e regulari quam li existent Europan',
    'Eine wunderbare Heiterkeit hat meine ganze Seele eingenommen, gleich den süßen Frühlingsmorgen, die ich mit ganzem Herzen genieße. Ich bin allein und freue mich meines Lebens in dieser Gegend, die für solche Seelen geschaffen ist wie die meine. Ich bin so glücklich, mein Bester, so ganz in dem Gefühle von ruhigem Dasein versunken, daß meine Kunst darunter leidet. Ich könnte jetzt nicht zeichnen, nicht einen Strich, und bin nie ein größerer Maler gewesen als in diesen Augenblicken. Wenn das liebe Tal um mich dampft, und die hohe Sonne an der Oberfläche der undurchdringlichen Finsternis meines Waldes ruht, und nur einzelne Strahlen sich in das innere Heiligtum stehlen, ich dann im hohen Grase am fallenden Bache liege, und näher an der Erde tausend mannigfaltige Gräschen mir merkwürdig werden; wenn ich das Wimmeln der kleinen Welt zwischen Halmen, die unzähligen, unergründlichen Gestalten der Würmchen, der Mückchen näher an meinem Herzen fühle, und fühle die Gegenwart des Allmächtigen, der uns nach seinem Bilde schuf, das Wehen des Alliebenden, der uns in ewiger Wonne schwebend trägt und erhält; mein Freund! Wenn\'s dann um meine Augen dämmert, und die Welt um mich her und der Himmel ganz in meiner Seele ruhn wie die Gestalt einer Geliebten - dann sehne ich mich oft und denke : ach könntest du das wieder ausdrücken, könntest du dem Papiere das einhauchen, was so voll, so warm in dir lebt, daß es würde der Spiegel deiner Seele, wie deine Seele ist der Spiegel des unendlichen Gottes! - mein Freund - aber ich gehe darüber zugrunde, ich erliege unter der Gewalt der Herrlichkeit dieser Erscheinungen.Eine wunderbare Heiterkeit hat meine ganze Seele eingenommen, gleich den süßen Frühlingsmorgen, die ich mit ganzem Herzen genieße. Ich bin allein und freue mich meines Lebens in dieser Gegend, die für solche Seelen geschaffen ist wie die meine. Ich bin so glücklich, mein Bester, so ganz in dem Gefühle von ruhigem Dasein versunken, daß meine Kunst darunter leidet. Ich könnte jetzt nicht zeichnen, nicht einen Strich, und bin nie ein größerer Maler gewesen als in diesen Augenblicken. Wenn das liebe Tal um mich dampft, und die hohe Sonne an der Oberfläche der undurchdringlichen Finsternis meines Waldes ruht, und nur einzelne Strahlen sich in das innere Heiligtum stehlen, ich dann im hohen Grase am fallenden Bache liege, und näher an der Erde tausend mannigfaltige Gräschen mir merkwürdig werden; wenn ich das Wimmeln der kleinen Welt zwischen Halmen, die unzähligen, unergründlichen Gestalten der Würmchen, der Mückchen näher an meinem Herzen fühle, und fühle die Gegenwart des Allmächtigen, der uns nach seinem Bilde schuf, das Wehen des Alliebenden, der uns in ewiger Wonne schwebend trägt und erhält; mein Freund! Wenn\'s dann um meine Augen dämmert, und die Welt um mich her und der Himmel ganz in meiner Seele ruhn wie die Gestalt einer Geliebten - dann sehne ich mich oft und denke : ach könntest du das wieder ausdrücken, könntest du dem Papiere das einhauchen, was so voll, so warm in dir lebt, daß es würde der Spiegel deiner',
]
y = [1, 2, 3]

vec = CountVectorizer()
pipeline = Pipeline(
    [('v', vec),
     ('c', GradientBoostingClassifier(n_estimators=5, max_depth=5,
                                      learning_rate=0.1,
                                      verbose=1, random_state=1234))])

x_train, x_test, y_train, y_test = train_test_split(data, y, random_state=1234)
pipeline.fit(x_train, y_train)

skl_to_pmml(pipeline, list(vec.get_feature_names()), 'y', '/tmp/reproduce.pmml')

```

**Comment by mariussoutier** [comment]
The script generates a PMML with an incorrect DataDictionary:

```
...
    <DataDictionary numberOfFields="231">
        <DataField name="aber" optype="categorical" dataType="string"/>
        <DataField name="ach" optype="continuous" dataType="double"/>
        <DataField name="al" optype="continuous" dataType="double"/>
        <DataField name="allein" optype="continuous" dataType="double"/>
...
```

**Comment by mariussoutier** [comment]
As far as I can see, that behavior is hard-coded in `pre_process.py':

```python
derived_colnames = get_derived_colnames('count_vec@[' + col_names[0] + ']', features)
```

**Comment by mariussoutier** [comment]
Not to overload this issue, but another thing I noticed is that in jpmml I also get this error when reading the model: `org.jpmml.evaluator.MissingValueException: The value for field "decisionFunction(0)" is not defined`.

**Comment by mariussoutier** [comment]
Are NumPy arrays even supported? It starts to get better when I use a DataFrame. Only exception left now is a problem with OutputFields in the Mining section.

**Comment by nyoka-pmml** [comment]
Below is the working example for your data:

```
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import train_test_split
import pandas as pd
from sklearn_pandas import DataFrameMapper
from nyoka import skl_to_pmml

data = [
    'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque eu, pretium quis, sem. Nulla consequat massa quis enim. Donec pede justo, fringilla vel, aliquet nec, vulputate eget, arcu. In enim justo, rhoncus ut, imperdiet a, venenatis vitae, justo. Nullam dictum felis eu pede mollis pretium. Integer tincidunt. Cras dapibus. Vivamus elementum semper nisi. Aenean vulputate eleifend tellus. Aenean leo ligula, porttitor eu, consequat vitae, eleifend ac, enim. Aliquam lorem ante, dapibus in, viverra quis, feugiat a, tellus. Phasellus viverra nulla ut metus varius laoreet. Quisque rutrum. Aenean imperdiet. Etiam ultricies nisi vel augue. Curabitur ullamcorper ultricies nisi. Nam eget dui. Etiam rhoncus. Maecenas tempus, tellus eget condimentum rhoncus, sem quam semper libero, sit amet adipiscing sem neque sed ipsum. Nam quam nunc, blandit vel, luctus pulvinar, hendrerit id, lorem. Maecenas nec odio et ante tincidunt tempus. Donec vitae sapien ut libero venenatis faucibus. Nullam quis ante. Etiam sit amet orci eget eros faucibus tincidunt. Duis leo. Sed fringilla mauris sit amet nibh. Donec sodales sagittis magna. Sed consequat, leo eget bibendum sodales, augue velit cursus nunc,',
    'Li Europan lingues es membres del sam familie. Lor separat existentie es un myth. Por scientie, musica, sport etc, litot Europa usa li sam vocabular. Li lingues differe solmen in li grammatica, li pronunciation e li plu commun vocabules. Omnicos directe al desirabilite de un nov lingua franca: On refusa continuar payar custosi traductores. At solmen va esser necessi far uniform grammatica, pronunciation e plu sommun paroles. Ma quande lingues coalesce, li grammatica del resultant lingue es plu simplic e regulari quam ti del coalescent lingues. Li nov lingua franca va esser plu simplic e regulari quam li existent Europan',
    'Eine wunderbare Heiterkeit hat meine ganze Seele eingenommen, gleich den süßen Frühlingsmorgen, die ich mit ganzem Herzen genieße. Ich bin allein und freue mich meines Lebens in dieser Gegend, die für solche Seelen geschaffen ist wie die meine. Ich bin so glücklich, mein Bester, so ganz in dem Gefühle von ruhigem Dasein versunken, daß meine Kunst darunter leidet. Ich könnte jetzt nicht zeichnen, nicht einen Strich, und bin nie ein größerer Maler gewesen als in diesen Augenblicken. Wenn das liebe Tal um mich dampft, und die hohe Sonne an der Oberfläche der undurchdringlichen Finsternis meines Waldes ruht, und nur einzelne Strahlen sich in das innere Heiligtum stehlen, ich dann im hohen Grase am fallenden Bache liege, und näher an der Erde tausend mannigfaltige Gräschen mir merkwürdig werden; wenn ich das Wimmeln der kleinen Welt zwischen Halmen, die unzähligen, unergründlichen Gestalten der Würmchen, der Mückchen näher an meinem Herzen fühle, und fühle die Gegenwart des Allmächtigen, der uns nach seinem Bilde schuf, das Wehen des Alliebenden, der uns in ewiger Wonne schwebend trägt und erhält; mein Freund! Wenn\'s dann um meine Augen dämmert, und die Welt um mich her und der Himmel ganz in meiner Seele ruhn wie die Gestalt einer Geliebten - dann sehne ich mich oft und denke : ach könntest du das wieder ausdrücken, könntest du dem Papiere das einhauchen, was so voll, so warm in dir lebt, daß es würde der Spiegel deiner Seele, wie deine Seele ist der Spiegel des unendlichen Gottes! - mein Freund - aber ich gehe darüber zugrunde, ich erliege unter der Gewalt der Herrlichkeit dieser Erscheinungen.Eine wunderbare Heiterkeit hat meine ganze Seele eingenommen, gleich den süßen Frühlingsmorgen, die ich mit ganzem Herzen genieße. Ich bin allein und freue mich meines Lebens in dieser Gegend, die für solche Seelen geschaffen ist wie die meine. Ich bin so glücklich, mein Bester, so ganz in dem Gefühle von ruhigem Dasein versunken, daß meine Kunst darunter leidet. Ich könnte jetzt nicht zeichnen, nicht einen Strich, und bin nie ein größerer Maler gewesen als in diesen Augenblicken. Wenn das liebe Tal um mich dampft, und die hohe Sonne an der Oberfläche der undurchdringlichen Finsternis meines Waldes ruht, und nur einzelne Strahlen sich in das innere Heiligtum stehlen, ich dann im hohen Grase am fallenden Bache liege, und näher an der Erde tausend mannigfaltige Gräschen mir merkwürdig werden; wenn ich das Wimmeln der kleinen Welt zwischen Halmen, die unzähligen, unergründlichen Gestalten der Würmchen, der Mückchen näher an meinem Herzen fühle, und fühle die Gegenwart des Allmächtigen, der uns nach seinem Bilde schuf, das Wehen des Alliebenden, der uns in ewiger Wonne schwebend trägt und erhält; mein Freund! Wenn\'s dann um meine Augen dämmert, und die Welt um mich her und der Himmel ganz in meiner Seele ruhn wie die Gestalt einer Geliebten - dann sehne ich mich oft und denke : ach könntest du das wieder ausdrücken, könntest du dem Papiere das einhauchen, was so voll, so warm in dir lebt, daß es würde der Spiegel deiner',
]

y = [1, 2, 3]

df = pd.DataFrame(data=data, columns=["x1"])

vec = CountVectorizer()
ppp = ('mapper', DataFrameMapper([
        ('x1', CountVectorizer())
    ]))
pipeline = Pipeline(
    [ppp,
     ('c', GradientBoostingClassifier(n_estimators=5, max_depth=5,
                                      learning_rate=0.1,
                                      verbose=1, random_state=1234))])

# x_train, x_test, y_train, y_test = train_test_split(df, y, random_state=1234)
pipeline.fit(df, y)

skl_to_pmml(pipeline, df.columns, 'y', 'reproduce.pmml')

```
We support numpy arrays if there are no pre-processings involved (It would be almost impossible to capture the new set of columns which are generated after pre-processings, (for example one-hot-encoding) and put it in the PMML).  Hope this answers your query.

**Comment by mariussoutier** [comment]
@nyoka-pmml Thanks a lot, I figured it only works with DataFrames. Now when I read in the model, there seems to be another problem that jpmml reports: `org.jpmml.evaluator.MissingAttributeException: Required attribute OutputField@dataType is not defined`, which is because of this line in the PMML:

```xml
        <Segmentation multipleModelMethod="modelChain">
            <Segment id="0" weight="1">
                <True/>
                <MiningModel modelName="MiningModel" functionName="regression">
                    <MiningSchema>
                        <MiningField name="document" usageType="active"/>
                    </MiningSchema>
                    <Output>
                        <OutputField name="decisionFunction(0)" feature="predictedValue" isFinalResult="false"/>
                        <OutputField name="transformedDecisionFunction(0)" feature="transformedValue">
                            <Apply function="+">
                                <Apply function="*">
                                    <FieldRef field="decisionFunction(0)"/>
                                    <Constant dataType="double">0.1</Constant>
                                </Apply>
                                <Constant dataType="double">0.0</Constant>
                            </Apply>
                        </OutputField>
                    </Output>
```

When I add `dataType="double"`, the error disappears.

**Comment by mariussoutier** [comment]
After filtering those fields out, I'm getting `org.jpmml.evaluator.DuplicateValueException: The value for field "probability_2" has already been defined`.

This part appears twice:

```xml
<Output>
                        <OutputField name="probability_2" optype="continuous" dataType="double" feature="probability" value="2"/>
                        <OutputField name="probability_3" optype="continuous" dataType="double" feature="probability" value="3"/>
                        <OutputField name="predicted_y" optype="categorical" dataType="string" feature="predictedValue"/>
                    </Output>
```

**Comment by Nirmal-Neel** [comment]
> 
> 
> @nyoka-pmml Thanks a lot, I figured it only works with DataFrames. Now when I read in the model, there seems to be another problem that jpmml reports: `org.jpmml.evaluator.MissingAttributeException: Required attribute OutputField@dataType is not defined`, which is because of this line in the PMML:
> 
> ```
>         <Segmentation multipleModelMethod="modelChain">
>             <Segment id="0" weight="1">
>                 <True/>
>                 <MiningModel modelName="MiningModel" functionName="regression">
>                     <MiningSchema>
>                         <MiningField name="document" usageType="active"/>
>                     </MiningSchema>
>                     <Output>
>                         <OutputField name="decisionFunction(0)" feature="predictedValue" isFinalResult="false"/>
>                         <OutputField name="transformedDecisionFunction(0)" feature="transformedValue">
>                             <Apply function="+">
>                                 <Apply function="*">
>                                     <FieldRef field="decisionFunction(0)"/>
>                                     <Constant dataType="double">0.1</Constant>
>                                 </Apply>
>                                 <Constant dataType="double">0.0</Constant>
>                             </Apply>
>                         </OutputField>
>                     </Output>
> ```
> 
> When I add `dataType="double"`, the error disappears.

Hi, this issue is solved in [Release Version 3.0.1](https://github.com/nyoka-pmml/nyoka/releases/tag/3.0.1)

**Comment by nyoka-pmml** [comment]
> 
> 
> And I'm also getting `org.jpmml.evaluator.DuplicateValueException: The value for field "probability_2" has already been defined`.
> 
> This part appears twice:
> 
> ```
> <Output>
>   <OutputField name="probability_2" optype="continuous" dataType="double" feature="probability" value="2"/>
>   <OutputField name="probability_3" optype="continuous" dataType="double" feature="probability" value="3"/>
>   <OutputField name="predicted_y" optype="categorical" dataType="string" feature="predictedValue"/>
> </Output>
> ```

This is resolved in [Release Version 3.1.0](https://github.com/nyoka-pmml/nyoka/releases/tag/3.1.0)

**Comment by nyoka-pmml** [comment]
[](url)
